warpd.1.md
warpd(1)

# DESCRIPTION

A modal keyboard driven pointer manipulation program.

# SYNOPSIS

warpd [options]

# OPTIONS

	*-f*, *--foreground*: Run warpd in the foreground. Mainly useful for debugging.

	*-l*, *--list-keys*: Print a list of valid keys which can be used as config values.

	*--list-options*: Print all configurable options.

	*-v*, *--version*: Print the current version.

	*-c*, *--config* <config file>: Use the provided config file (- corresponds to stdin).

Mode Flags:

	*--hint*: Run warpd in (daemonless) hint mode.

	*--hint2*: Run warpd in 2 stage hint mode.

	*--grid*: Run warpd in grid mode.

	*--normal*: Run warpd in normal mode.

	*--history*: Run warpd in history hint mode.

	*--screen*: Run warpd in screen selection mode.

	*--oneshot*: When paired with one of the mode flags, exit warpd as soon as the mode is complete (i.e don't drop into normal mode). When used with normal mode, exit as soon as a mouse button is pressed (without pressing it). If no mode flag is specified, the default behaviour is the same as --normal --oneshot. Principally useful for scripting.

	*-q*, *--query*: Consume a list of hints from stdin and print the result to stdout (in the form <x> <y> <hint>). Each line should have the form _<label> <x> <y>_. May be used in conjunction with --click.

	*--drag*: Automatically start a drag operation when paired with --normal.

	*--move '<x> <y>'*: Move the pointer to the specified coordinates and exit.

	*--click <button>*: Send a mouse click corresponding to the supplied button and exit. May be paired with --move or --oneshot (in which case the click will occur at the end of the selection).

	*--record*: When used with --click, adds the click event to warpd's history.

# DESCRIPTION

warpd has three main modes which can be used to manipulate the pointer. The
primary mode is called 'normal mode' (A-M-c) and facilitates local pointer
movement using vi-like bindings (_h_ _j_ _k_ _l_). The other two modes, *hint*
and *grid* mode are used to effect larger movements across the screen and are
expected to be used in combination with normal mode to achieve the desired end.

For example, the user might activate warpd in hint mode (_A-M-x_) to pinpoint
the start of a text region before starting a drag operation (_v_) and
ultimately using normal mode to complete the selection. If the selection is a
large body of text, it may be desirable to activate grid (_g_) or hint (_x_)
mode for a second time to warp the pointer to the desired region's terminal
point.  

In its simplest form, warpd can be used as a poor man's xdotool (--move/--click).
Scrolling can be acheived by using _e_ and _y_ in normal mode. For a full list of
bindings see _--list-options_.

A description of each mode follows (see also _USAGE_NOTES_):

## Normal Mode (A-M-c)

This is the default mode (and the endpoint of both grid and normal mode) which
is designed for short distance pointer manipulation. It is particularly useful
for manipulating popup menus and selecting text (see _Dragging_). The default
behaviour is vi-like. Pressing the mapped directional keys (default hjkl) moves
the cursor in a continuous fashion, but the pointer can also be warped to the edges
of the screen using the home (_H_), middle (_M_), and last (_L_) mappings.
Finally, a numeric multiplier can be supplied to the
directional keys as an input prefix in order to move the cursor by a
proportional increment in the given direction (e.g 10j moves 10 units down). 

## Hint Mode (A-M-x or simply 'x' within normal mode)

This mode populates the screen with a list of labels and allows the
user to immediately warp the pointer to a given location by pressing the
corresponding key sequence. It is similar to functionality provided by browser
plugins like Vimperator but works outside of the browser and indiscriminately
covers the entire screen. Once a target has been selected 'normal mode' is
entered for further manipulation.

*Note:* While it may at first be tempting to saturate the screen with hints,
the user is cautioned against this. A balance must be struck between hint size,
the number of hints, and the size of the screen. Enough space must be left to
provide contextual awareness, while enough hints must be present to facilitate
targetting UI elements without the need for too much adjustment. Once this
equilibrium has been achieved, using hint mode become second nature and is (in
the author's opinion) superior to grid mode for quickly pinpointing elements.

For finer movements, a two phase hint mode can be activated by pressing 'X'
within normal mode.

## History Mode (';' within normal mode)

Identical to hint mode but exclusively displays hints over previously
selected targets.

## Grid Mode ('A-M-g' or simply 'g' within normal mode)

By default grid mode divides the screen into a 2x2 grid. Each time a key
is pressed the grid shrinks to cover the targeted quadrant. The cursor is placed
in the middle of the grid.

covers the desired target a mouse button (e.g 'm') can be pressed.


E.G

```
         +--------+--------+            +--------+--------+
         |        |        |            |  u |  i |       |
         |   u    |   i    |            |----m----+       |
 M-x     |        |        |     u      |  j |  k |       |
----->   +--------m--------+   ----->   +---------+       |
         |        |        |            |                 |
         |   j    |   k    |            |                 |
         |        |        |            |                 |
         +--------+--------+            +--------+--------+
```

## Screen Selection Mode ('A-M-s' or simply 's' within normal mode)

This mode is intended for multi-screen setups and provides the user with a
dedicated set of hints for switching monitors and dropping them into normal
mode.

## Dragging

Pressing _v_ whilst in normal mode toggles a drag operation. The cursor can
then be warped around the screen as normal in order to select text or move
objects until the drag key is hit again or a mouse button is pressed.
Additionally, the *copy_and_exit* key (_c_) may be used to copy the selected
text to the system clipboard and terminate the current session.

## Wayland

Wayland's architecture does not allow clients to globally bind hotkeys. In
order to run warpd, one of the mode flags must be used. These can be bound
to the corresponding hotkeys in your compositor.

E.G

On sway:

```
bindsym Mod4+Mod1+x exec warpd --hint
bindsym Mod4+Mod1+c exec warpd --normal
bindsym Mod4+Mod1+g exec warpd --grid
```

Non-wayland users should favour the daemon, since it also caches some of the
draw operations to improve performance.

# CONFIG OPTIONS

The program can be modified by placing configuration options in 
*~/.config/warpd/config*, a complete list of which can be obtained
with _--list-options_.

Each option must be specified on its own line and have the format:

`<option>: <value>`

Options which take one or more keys may optionally be unset using the value
"unbind".

## RULES

- Options which expect one or more keys may be specified multiple times, in which case all accepted mappings are interchangeable. 

- Options which accept multiple keys (e.g _buttons_) expect each key to be separated by a space.

- For options expecting only a single key, it is possible to specify all desired bindings as space separated values in a single declaration.

- If a key is bound to multiple config options, the most recently defined one takes precedence.

## E.G

```
hint_mode: f a
```

is identical to

```
hint_mode: f
hint_mode: a
```

and will bind both *f* and *a* to hint mode activation keys.

while

```
buttons: m , .
buttons: j k l
```

will bind both the sequences *m* *,* *.* and *j* *k* *l* to mouse buttons *1* *2*
and *3* respectively. Note that the latter set of bindings override the default
normal mode movements keys.

# SCRIPTING

Judicious use of warpd's flags makes creating custom modes fairly easy.

For example, the --oneshot flag may be used in conjunction with one of the mode
flags to facilitate target selection at which point subsequent action can be
taken by the script.

## Examples

	# Prompt the user for a target using two stage hint mode and 
	# left click on the result. 
	#
	# This particular example can be achieved more concisely with:
	#
	# warpd --hint2 --click 1

	warpd --hint2 --oneshot; warpd --click 1

	# Interactively create a hint file.
	#
	# Points can be added by navigating around as usual and pressing 'p'.
	# The process is terminated by using the exit key.

	warpd --normal|awk '{print substr("asdfghjkl;qwertyuiopzxcv,./;", NR, 1), $0}' > hints

	# Prompt the user for one of the recorded hints and click on the
	# result.

	warpd --click 1 --query < hints 


# USAGE NOTES

The key to using warpd effectively is to learn when to exit normal mode. Much
of one's time at a computer is spent moving the mouse between windows,
interacting with UI elements, and reading text. What one might call 'browse
mode'. It is in this mode of operation that it makes sense to keep warpd
active.

Developing facility with the scroll and oneshot mouse buttons is key to
achieving this. For example, if you happen to have two documents open and wish
to switch between them, you can simply type _x fx_ (where _fx_ is a hint) if
normal mode is already active. Scrolling can subsequently be achieved using _e_ and
_r_. Once you finally wish to type something, you can do _x fx n_ to focus on
the UI element, click, and exit.

Conversely, warpd can complement an input heavy workflow with its oneshot
functionality and dedicated activation keys (E.G _n_, _A-M-l_, _A-M-x_, etc).  

It is important to note that warpd is not intended to replace mouse heavy
workflows, and will likely always be inferior for rapid precise local
movements. When confronted with an IDE, or some other pointer driven
abomination, the author still sometimes reaches for his mouse.

## On Dragging

Activating discrete mode and pressing v can provide a familiar environment to a
_vi_ user but it is important to remember that cursor manipulation is
application agnostic and consequently ignorant of the text on the screen. All
movement is necessarily pixel based, consequently, drag + hint
mode can be a superior method for surgically selecting text (though it may at
first be less intuitive).

# FILES

*$XDG_CONFIG_HOME/warpd/config*++
*~/.config/warpd/config*
	The path of the configuration file (searched in order).

# EXIT STATUS

On error warpd will exit with a status of 255, otherwise the exit code will
correspond to the button used to terminate warpd (in the case of oneshot
buttons or in the presence of --oneshot).  If warpd is terminated explicitly
(i.e with the exit key) it will exit with a status of 0.

# BUGS/LIMITATIONS

warpd uses various platform specific hacks to bypass limitations
of the display server. All implementations were written by the
same author, who presently uses X.

Consequently testing on non-X platforms has been minimal.

YMMV

Bugs can be reported here:

https://github.com/rvaiya/warpd/issues/

A list of known limitations follow:

- Multiscreen support currently does not support hotplugging. This means that
  you must restart warpd after making any changes to your screen configuration.

- For implementation reasons, the cursor position is not horizontally centered,
  but to the right of the actual pointer. This generally isn't an issue,
  but may become more noticeable as you increase _cursor_size_.

## X

- Unplugging the keyboard while warpd is one of its active modes will cause
  pandemonium.  If you do this (don't :P), you may need to remotely ssh into
  the machine or switch to a VT to kill the process.

- warpd uses Xinput for input processing to bypass certain limitation of the X
  input system. A byproduct of this is that certain remapping tools will not
  work (e.g xcape). If you are in the habit of making unorthodox changes to
  your keymap (like remapping capslock to control/escape) you may want to try
  an evdev based remapper like keyd (https://github.com/rvaiya/keyd).

- Programs which use Xinput to directly manipulate input devices may misbehave.
  See [Issue #3](https://github.com/rvaiya/warpd/issues/3#issuecomment-628936249) for details.

## MacOS

- Cursor hiding relies on a hack that some programs ignore (e.g iTerm).
- Some programs (e.g iTerm) have a 'secure input mode' (which can usually be
disabled) that causes interference.

## Wayland

- Cursor hiding doesn't work.
- Running as a daemon doesn't work (can't listen for hotkeys).
- UI elements (e.g input fields) which require focus can't be selected.

# AUTHORS

Written and maintained by Raheman Vaiya (2019-).
See https://github.com/rvaiya/warpd for more information.


CHANGELOG.md
# v1.3.5
- macos: Various input bugfixes
- macos: Introduce launchd service (the user should no longer run warpd explicitly)
- macos: Fix multi-user support
- linux: Merge X/Wayland support into a single binary
- Improve portrait mode support
- Make the drag button configurable
- Fix config key precedence bug

# v1.3.4
 - Fix initial pointer position on wayland
 - Add support middle click support for macos
 - Add support for platform keymaps (non-qwerty layouts)
 - Various bugfixes

# v1.3.3
 - Fix wayland initialization issue

# v1.3.2
 - Improve config handling (allow for shadowed values to gracefully handle conflicts)
 - Make undo keys in hint based modes configurable

# v1.3.1

 - Allow secondary mode selection with --oneshot --normal.
 - Add identifying information to print keys and -q
 - Fix -c

# v1.3.0

 - Add more scripting facilities (e.g --oneshot --record)
 - Allow oneshot activation while the daemon is running
 - Fix -q flag
 - Add print key
 - Make hint size specification consistent
 - Add --query
 - Fix dialog occlusion issue (#134)
 - Implement 2 stage hint mode
 - Add history mode
 - Fix modified button presses (#86)
 - Fix macos compilation issues
 - Add an optional visual indicator for normal mode (#121)
 - Check for zxdg_output_manager before adding screens (#116)
 - Make scroll speed configurable (#112)
 - wayland: Make font configurable (#137)
 - wayland: Fix multi-screen support (#135)
 - Various bugfixes

# v1.2.2

 - Use XDG compliant config paths + eliminate ~/.config pollution
 - wayland: Lower the required layer_shell version
 - Eliminate hint asymmetry (alters the unit of hint_size)
 - Add dedicated `hint_exit` key (#99)
 - Add --config, and --help
 - Misc bugfixes

# v1.2.1

 - Add multi-screen support to wayland
 - Add dedicated hint_exit key (distinct from exit)

# v1.2.0

 - Add experimental wayland support
 - Add --hint/--normal/--grid flags
 - Fix missing key up bug

# v1.1.4-beta

 - Fix alternate (english) layout support

# v1.1.3-beta

 - Hide cursor when scrolling
 - Fix full hint drawing issue caused by switching VTs

# v1.1.2-beta

 - Add support for modified button presses
 - Add accelerator key

# v1.1.1-beta

 - Add support for shifted hint characters
 - Fix screen boundary issue
 - Limit maximum hint size for multi-screen/resolution setups
 - Change the config list delimiter from comma to space
 - Update man page
 - Misc bugfixes

# v1.1.0-beta

 - Add Multiscreen support
 - Make the default colour scheme less appalling
 - Make other modes accessible from grid mode

# v1.0.3-beta

 - macos: solicit accessibility whitelisting on initialization (#64)
 - macos: cleanup input handling code + fix freezing (#65)
 - Account for different X maps

# v1.0.2-beta

 - macos: Fix sticky key visual indicators
 - Add pointer acceleration (`acceleration`)
 - Add `grid_border_size` and `grid_border_color`
 - Add `hint_oneshot_key`
 - Show cursor in grid mode

# v1.0.0-beta

 - Add MacOS support
 - Internal code cleanup
 - Normal movement keys are now continuous rather than discrete.
 - Remove 'normal_' prefix from normal options.
 - Eliminate hint opacity variable in favour of rgba hex values.
 - Eliminate scroll flinging

# v0.0.1

In the beginning there was only darkness.



src/warpd.c
/*
 * warpd - A modal keyboard-driven pointing system.
 *
 * © 2019 Raheman Vaiya (see: LICENSE).
 */

#include "warpd.h"

struct platform *platform = NULL;

static const char *config_path;

uint64_t get_time_us()
{
	struct timespec ts;
	clock_gettime(CLOCK_MONOTONIC, &ts);

	return ts.tv_nsec / 1E3 + ts.tv_sec * 1E6;
}


const char *get_data_path(const char *file)
{
	static char path[PATH_MAX];

	if (getenv("XDG_DATA_DIR")) {
		sprintf(path, "%s/warpd", getenv("XDG_DATA_DIR"));
		mkdir(path, 0700);
	} else {
		sprintf(path, "%s/.local", getenv("HOME"));
		mkdir(path, 0700);
		strcat(path, "/share");
		mkdir(path, 0700);
		strcat(path, "/warpd");
		mkdir(path, 0700);
	}

	strcat(path, "/");
	strcat(path, file);

	return path;
}

const char *get_config_path(const char *file)
{
	static char path[PATH_MAX];

	if (getenv("XDG_CONFIG_HOME")) {
		sprintf(path, "%s/warpd", getenv("XDG_CONFIG_HOME"));
		mkdir(path, 0700);
	} else {
		sprintf(path, "%s/.config", getenv("HOME"));
		mkdir(path, 0700);
		strcat(path, "/warpd");
		mkdir(path, 0700);
	}

	strcat(path, "/");
	strcat(path, file);

	return path;
}


static void lock()
{
	int fd;
	char path[64];

	sprintf(path, "/tmp/warpd_%d.lock", getuid());
	fd = open(path, O_RDONLY|O_CREAT, 0600);

	if (fd < 0) {
		perror("flock open");
		exit(-1);
	}

	if (flock(fd, LOCK_EX | LOCK_NB) == -1) {
		fprintf(
		    stderr,
		    "ERROR: Another instance of warpd is already running.\n");
		exit(-1);
	}
}

static void daemonize()
{
	if (fork())
		exit(0);
	if (fork())
		exit(0);

	int fd = open("/dev/null", O_WRONLY);
	if (fd < 0) {
		perror("open");
		exit(-1);
	}

	close(1);
	close(2);

	dup2(fd, 1);
	dup2(fd, 2);
}

static void print_usage()
{
	const char *usage =
		"warpd: [options]\n\n"
		"  -f, --foreground            Run warpd in the foreground (useful for debugging).\n"
		"  -h, --help                  Print this help message.\n"
		"  -v, --version               Print the version and exit.\n"
		"  -c, --config <config file>  Use the supplied config file.\n"
		"  -l, --list-keys             Print all valid keys.\n"
		"  --list-options              Print all available config options.\n"

		"  --hint                      Start warpd in hint mode and exit after the end of the session.\n"
		"  --hint2                     Start warpd in two pass hint mode and exit after the end of the session.\n"
		"  --normal                    Start warpd in normal mode and exit after the end of the session.\n"
		"  --grid                      Start warpd in hint grid and exit after the end of the session.\n"
		"  --screen                    Start warpd in screen selection mode and exit after the end of the session.\n"
		"  --oneshot                   When paired with one of the mode flags, exit warpd as soon as the mode is complete (i.e don't drop into normal mode). Principally useful for scripting.\n"
		"  --move '<x> <y>'            Move the pointer to the specified coordinates.\n"
		"  --click <button>            Send a mouse click corresponding to the supplied button and exit. May be paired with --move.\n"
		"  -q, --query                 Consumes a list of hints from stdin and presents a one off hint selection.\n"
		"  --record                    When used with --click, records the event in warpd's hint history.\n\n"
		;

	printf("%s", usage);
}

static void print_version()
{
	printf("warpd " VERSION"\n");
}


static int drag_flag = 0;
static int oneshot_flag = 0;
static int click_flag = 0;
static int x_flag = -1;
static int y_flag = -1;
static int record_flag = 0;
static int mode = 0;

/* Platform entry points. */
int oneshot_main(struct platform *_platform)
{
	int ret = 0;
	screen_t scr;
	platform = _platform;

	parse_config(config_path);
	init_mouse();
	init_hints();

	platform->mouse_get_position(&scr, NULL, NULL);
	if (x_flag == -1 && y_flag == -1) {
		if (drag_flag)
			platform->mouse_down(config_get_int("drag_button"));

		ret = mode_loop(mode, oneshot_flag, record_flag);

		if (drag_flag)
			platform->mouse_up(config_get_int("drag_button"));

	} else {
		platform->mouse_move(scr, x_flag, y_flag);
	}

	if (click_flag)
		platform->mouse_click(click_flag);

	return ret;
}

int daemon_main(struct platform *_platform)
{
	platform = _platform;

	parse_config(config_path);
	init_mouse();
	init_hints();

	daemon_loop(config_path);

	return 0;
}

int print_keys_main(struct platform *platform)
{
	size_t i;
	for (i = 1; i < 256; i++) {
		const char *name = platform->input_lookup_name(i, 0);

		if (name && name[0])
			printf("%s\n", name);

		name = platform->input_lookup_name(i, 1);
		if (name && name[0])
			printf("%s\n", name);
	}

	return 0;
}

int main(int argc, char *argv[])
{
	int c;
	int foreground = 0;
	config_path = get_config_path("config");

	struct option opts[] = {
		{"version", no_argument, NULL, 'v'},
		{"help", no_argument, NULL, 'h'},
		{"query", no_argument, NULL, 'q'},
		{"list-keys", no_argument, NULL, 'l'},
		{"foreground", no_argument, NULL, 'f'},
		{"config", required_argument, NULL, 'c'},

		{"hint", no_argument, NULL, 257},
		{"grid", no_argument, NULL, 258},
		{"normal", no_argument, NULL, 259},
		{"hint2", no_argument, NULL, 261},
		{"history", no_argument, NULL, 262},
		{"list-options", no_argument, NULL, 260},
		{"oneshot", no_argument, NULL, 263},
		{"click", required_argument, NULL, 264},
		{"move", required_argument, NULL, 265},
		{"record", no_argument, NULL, 266},
		{"drag", no_argument, NULL, 267},
		{"screen", no_argument, NULL, 268},
		{0}
	};

	while ((c = getopt_long(argc, argv, "qrhfvlc:", opts, NULL)) != -1) {
		switch (c) {
			case 'v':
				print_version();
				return 0;
			case 'h':
				print_usage();
				return 0;
			case 'l':
				platform_run(print_keys_main);
				return 0;
			case 'c':
				config_path = optarg;
				break;
			case 'f':
				foreground = 1;
				break;
			case 'q':
				mode = MODE_HINTSPEC;
				oneshot_flag = 1;
				break;
			case 257:
				mode = MODE_HINT;
				break;
			case 258:
				mode = MODE_GRID;
				break;
			case 259:
				mode = MODE_NORMAL;
				break;
			case 261:
				mode = MODE_HINT2;
				break;
			case 262:
				mode = MODE_HISTORY;
				break;
			case 268:
				mode = MODE_SCREEN_SELECTION;
				break;
			case 263:
				if (!mode)
					mode = MODE_NORMAL;

				oneshot_flag = 1;
				break;
			case 264:
				click_flag = atoi(optarg);
				oneshot_flag = 1;
				break;
			case 265:
				sscanf(optarg, "%d %d", &x_flag, &y_flag);
				oneshot_flag = 1;
				break;
			case 266:
				record_flag = 1;
				break;
			case 267:
				drag_flag = 1;
				break;
			case 260:
				config_print_options();
				return 0;
			case '?':
				return -1;
		}
	}

	if (mode || oneshot_flag) {
		platform_run(oneshot_main);
	} else {
		lock();

		if (!foreground)
			daemonize();

		setvbuf(stdout, NULL, _IOLBF, 0);
		printf("Starting warpd " VERSION "\n");

		platform_run(daemon_main);
	}
}
